src/ExperimentTable.js

```diff
@@ -103,33 +131,50 @@ class ExperimentTable extends React.Component {
   }
 
   render() {
-    const { searchQuery, searchedColumnIndex, selectedSearch, selectedKingdom, checkedRows } = this.state
+    const { searchQuery, searchedColumnIndex, selectedSearch, selectedDropdownFilters, checkedRows } = this.state
     const { orderedColumnIndex, ascendingOrder } = this.state
     const { entriesPerPage, currentPage } = this.state
-    const { host, aaData, tableHeader, enableDownload, downloadTooltip } = this.props
+    const { host, aaData, tableHeader, enableDownload, downloadTooltip, tableFilters } = this.props
 
     const displayedFields = tableHeader.map(header => header.dataParam)
 
-    const dataArray = selectedSearch.trim() ?
+    const experimentTableFilters = tableFilters.map(filter => {
+      return {
+        label: filter.label,
+        options: _.uniq(
+          _.flatMap(aaData.map(data => data[filter.dataParam])))
+      }
+    })
+
+    const tableHeaderFilteredExperiments = this.filter(this.sort(aaData), tableHeader)
+
+    const selectedSearchFilteredExperiments = this.sort(aaData).filter(data => data &&
+      Object.keys(data).map(key => displayedFields.includes(key) ? data[key] : null)
+        .some(value => value && value.toString().toLowerCase().includes(selectedSearch)))
+
+    const dropdownFilteredExperiments = selectedDropdownFilters.length === 0 ?
+      tableHeaderFilteredExperiments : tableHeaderFilteredExperiments.filter(data => {
+        return selectedDropdownFilters.every(filter => {
+          return filter ? this._isContains(data, filter.value) : true
+        })
+      })
 
-      this.sort(aaData).filter(data => data &&
-        Object.keys(data).map(key => displayedFields.includes(key) ? data[key] : null)
-          .some(value => value && value.toString().toLowerCase().includes(selectedSearch))) :
-      this.filter(this.sort(aaData), tableHeader)
-        .filter(data => selectedKingdom ? data.kingdom === selectedKingdom : true)
+    const dataArray = selectedSearch.trim() ?
+      _.intersection(selectedSearchFilteredExperiments, dropdownFilteredExperiments) :
+      dropdownFilteredExperiments
```

Comment:
Think about the logic here: you’re filtering the whole set of results by two different set of criteria, and then doing the interesection of the two result sets. It’s way more efficient and concise filtering by the first set of criteria, and then chain the second `filter` to the remaining elements.

Reply:
Fixed